flink-clients/src/test/java/org/apache/flink/client/program/PackagedProgramRetrieverImplTest.java

```diff
@@ -0,0 +1,590 @@
+
+        // setting no jar, user lib dir and jobClassName makes sure that everything's picked up from
+        // the system classpath
+        final PackagedProgramRetriever retriever =
+                PackagedProgramRetrieverImpl.create(
+                        null, null, null, new String[0], new Configuration());
+        assertThat(
+                retriever.getPackagedProgram().getMainClassName(),
+                is(classpathProvider.getJobClassName()));
+    }
+
+    @Test
+    public void testJobGraphRetrieval()
+            throws IOException, FlinkException, ProgramInvocationException {
+        final int parallelism = 42;
+        final JobID jobId = new JobID();
+
+        final Configuration configuration = new Configuration();
+        configuration.setInteger(CoreOptions.DEFAULT_PARALLELISM, parallelism);
+        configuration.set(PipelineOptionsInternal.PIPELINE_FIXED_JOB_ID, jobId.toHexString());
+
+        classpathProvider.setSystemClasspathWithEntryClass();
+
+        final String expectedSuffix = "suffix";
+        final PackagedProgramRetriever retriever =
+                PackagedProgramRetrieverImpl.create(
+                        null,
+                        classpathProvider.getJobClassName(),
+                        classpathProvider.getTestJobArgs(expectedSuffix),
+                        new Configuration());
+
+        final JobGraph jobGraph = retrieveJobGraph(retriever, configuration);
+
+        assertThat(
+                jobGraph.getName(), is(classpathProvider.getJobClassName() + "-" + expectedSuffix));
+        assertThat(jobGraph.getSavepointRestoreSettings(), is(SavepointRestoreSettings.none()));
+        assertThat(jobGraph.getMaximumParallelism(), is(parallelism));
+        assertThat(jobGraph.getJobID(), is(jobId));
+    }
+
+    @Test
+    public void testJobGraphRetrievalFromJar()
+            throws IOException, FlinkException, ProgramInvocationException {
+        final String expectedSuffix = "suffix";
+        final PackagedProgramRetriever retrieverUnderTest =
+                PackagedProgramRetrieverImpl.create(
+                        classpathProvider.getUserDirectoryWithTestJob(),
+                        null,
+                        null,
+                        classpathProvider.getTestJobArgs(expectedSuffix),
+                        new Configuration());
+
+        final JobGraph jobGraph = retrieveJobGraph(retrieverUnderTest, new Configuration());
+
+        assertThat(
+                jobGraph.getName(),
+                is(classpathProvider.getTestJobClassName() + "-" + expectedSuffix));
+    }
+
+    @Test
+    public void testJobGraphRetrievalJobClassNameHasPrecedenceOverClasspath()
+            throws IOException, FlinkException, ProgramInvocationException {
+        // TODO: I don't get this test: Shouldn't the related jar be on the classpath?
+        //        classpathProvider.setSystemClasspathWithTestJob();
```

Comment:
This test was migrated from the old `ClassPathPackagedProgramRetrieverTest`. But it feels to me that we don't want that behavior, do we?